### ISSUE TITLE
Version numbers starting with 0 cause problems with vulnerability detection

### ISSUE THREAD

**Issue by JorritSalverda** [kubernetes/client-go#969]
Vulnerability *CVE-2019-11250* fixed in `1.15.4` and `1.16.1-beta.0` reappears due to the version numbers having drop to 0.x.x. 

This happens in the `trivy` scanner, see issue https://github.com/aquasecurity/trivy/issues/985#issuecomment-845896739

It outputs the following:

```
+------------------+------------------+----------+-------------------+-----------------------+
|     LIBRARY      | VULNERABILITY ID | SEVERITY | INSTALLED VERSION |     FIXED VERSION     |
+------------------+------------------+----------+-------------------+-----------------------+
| k8s.io/client-go | CVE-2019-11250   | MEDIUM   | v0.21.1           | 1.15.4, 1.16.1-beta.0 |
+------------------+------------------+----------+-------------------+-----------------------+
```

Based on the version numbers it's quite impossible for any vulnerability scanner to see this as fixed in the `0.x.x` version numbers. Any suggestion on how this can be resolved? Otherwise this will continue to pop up as a vulnerability until the version numbers jump beyond 1.16.

**Comment by liggitt** [comment]
> fixed in 1.15.4 and 1.16.1-beta.0

Those version numbers aren't valid for published libraries like client-go or apimachinery, so they should not be used in an automated advisory connected to those libraries, but should be translated from a Kubernetes version to the library versions.

**Comment by liggitt** [comment]
What originated the k8s.io/client-go association with that CVE?

**Comment by knqyf263** [comment]
Thanks for your reply, @liggitt. As you said, there aren't v1.15.4 and v1.16.1-beta.0 in client-go, so I think the following advisory is wrong.
https://gitlab.com/gitlab-org/security-products/gemnasium-db/-/blob/93c31600e2a0f9ec4db2e584d1781f3cf2e28685/go/k8s.io/client-go/CVE-2019-11250.yml#L13-14

On the other hand, the Go Vulnerability Database says the fixed version is v0.17.0.
https://github.com/golang/vulndb/blob/e0c00fae09e687ec6febda47ae3bc7552fc7b988/reports/GO-2021-0065.yaml#L11

The fixed commit seems to be included in Kubernetes v1.16.0.
https://github.com/kubernetes/kubernetes/commit/4441f1d9c3e94d9a3d93b4f184a591cab02a5245

I think the following commit fixed the vulnerability in client-go. Which published version fixes the vulnerability in client-go? v0.17.0?
https://github.com/kubernetes/client-go/commit/1ca6f0db2b22a5888d39562ebf3745bb035a6a09

I'm confused with `kubernetes-1.16.0` tag saying "We currently recommend using the v12.0.0 tag".
https://github.com/kubernetes/client-go/tree/kubernetes-1.16.0

Is this originally fixed in Kubernetes 1.16.0, but in client-go it was fixed in 0.17.0? If so, I'll ask GitLab to fix the advisory.

**Comment by liggitt** [comment]
> I'm confused with kubernetes-1.16.0 tag saying "We currently recommend using the v12.0.0 tag".
https://github.com/kubernetes/client-go/tree/kubernetes-1.16.0

That is confusing. I'm sorry. 

The guidance to use v12.0.0 was flawed, and was based on the former approach of using major version tags per client-go release. That stopped working when we transitioned to go modules in Kubernetes 1.16.

The fix is available in the non-semver kubernetes-1.16.0 tag (https://github.com/kubernetes/client-go/commits/kubernetes-1.16.0/transport/round_trippers.go)

We also started publishing semver v0.x.y tags corresponding to Kubernetes v1.x.y once the issue with major version tags was recognized, so the fix is also available in client-go v0.16.4+

Using v0.x.y client-go tags is the recommendation for current versions

**Comment by knqyf263** [comment]
I understand. If the security advisory changes the fixed version to 0.16.4, the old version with the former approach such as v12.0.0 will appear to be unaffected by the vulnerability since v12.0.0 is newer than v0.16.4 in semver. I thought it was problematic.

However, v12.0.0 isn't present in the go module world if I understand correctly. `kubernetes-1.16.0` tag is converted into `v0.0.0-20190918160344-1fbdaa4c8d90`. This is older than v0.16.4 and doesn't cause any issues in vulnerability detection.

```
$ go get k8s.io/client-go@kubernetes-1.16.0
go get: added k8s.io/client-go v0.0.0-20190918160344-1fbdaa4c8d90
```

In short, it would be ok for users who use client-go with go modules once we change the fixed version in the advisory to v0.16.4. Please correct me if I'm wrong. Thanks for your explanation!

**Comment by liggitt** [comment]
> In short, it would be ok for users who use client-go with go modules once we change the fixed version in the advisory to v0.16.4. Please correct me if I'm wrong.

for go module users, that's correct

**Comment by knqyf263** [comment]
Thanks for your confirmation.
